### PR TITLE
Fix 1349

### PR DESCRIPTION
**Pull request by dsyme**
Fix for https://github.com/Microsoft/visualfsharp/issues/1349

For the InnerLambdasToTopLevelFuncs (lambda lifting) optimization, all variables reported as free by ``accFreeInExpr`` must also be visited and logged by ``FoldExpr``.  

In the cases of a quotation, this wasn't happening, causing an error if a variable with constant value was optimized away and no longer appeared in the spliced expressions of the quotation . The  root cause is that the ``FoldExpr`` was not visiting the body of the quotation, but the ``accFreeInExpr`` was.

### CODE REVIEW

File: tests/fsharp/core/quotes/test.fsx [hunk 1]
```diff
@@ -2718,6 +2718,15 @@ module ExtensionMembersWithSameName =
...
 #endif
 
+module PartialApplicationLeadToInvalidCodeWhenOptimized = 
+    let f () = 
+        let x = 1
+        let g (y:int) (z:int) = <@ x @>
+        let _ = g 3 // the closure generated by this code was invalid
+        ()
+
+    f ()
```
Comment (forki) [inline tests/fsharp/core/quotes/test.fsx]: What is the assertion here? Is this code run? Do we run it through PeVerify?

Comment (dsyme) [inline tests/fsharp/core/quotes/test.fsx]: Yes, we run it through peverify.

**Comment by eriawan** [comment]
@dsyme 
Pardon, any chance to have this fix available (to be downloaded) before the RTM of Visual Studio "15"?

**Comment by dsyme** [comment]
@eriawan No, I don't really think so.  There's sure to be a workaround in any specific case - the problem was quite corner-case.  For example, avoiding the use of local curried functions and instead defining them at the top level of your program.

**Pull request merged**